Performance Investigation: Gunicorn async and threaded web servers (#69)
ONSdigital/eq-survey-runner-benchmark: doc/performance-investigations/0010-web-server-worker-class/summary.md

> # Web server worker class: Async vs Threads vs Threads + GRPC

Reviewer: Isn't it gRPC?

> # Web server worker class: Async vs Threads vs Threads + gRPC

Reviewer 2: We have previously tested async vs threads vs threads + gRPC in https://github.com/ONSdigital/eq-survey-runner-benchmark/blob/master/doc/performance-investigations/0008-use-gthread-grpc/summary.md. I think we should present this as a re-run of that investigation now that sessions are in Datastore. By linking to that investigation all of the documentation around the why of threads and gRPC is already covered.

Author: Done.

> We have only ever run the application with Gunicorn in async mode (workers). We should try threads if it is more performant, reliable or consistent. 

Reviewer: Past doesn't really mean much or provide a justification for a test, specially if you are reading in the future, as context is lost . This is simply a Gunicorn in async mode compared to it using threads with and without gRPC

Author: Just following our past pattern. But have updated.

 - the ability to use other Google libraries such as Pub/Sub which current is only performant under threads

## Benchmark profile

| Option                 | Value                        |
|------------------------|------------------------------|
| Requests file          | census_household_gb_eng.json |
| Run time               | 30m                          |
| User wait time minimum | 1                            |
| User wait time maximum | 2                            |
| Clients                | 64                           |
| Hatch rate             | 64                           |
| Number of workers      | 7                            |
| Number of threads      | 7                            |

This test replicates the current settings for the daily test.

The version of runner tested: https://github.com/ONSdigital/eq-questionnaire-runner/releases/tag/v3.53.0

## Results

The tests were run multiple times to ensure the performance was consistent; however, I have only included a single sample per configuration in the outputs.
The results below are the averages between all tests. Note response times are in milliseconds.

> |                          | Request Count | Failure Count | Median Response Time | Average Response Time | Min Response Time | Max Response Time | Average Content Size | Requests/s |

Reviewer: With my maths hat on, you shouldn't mix median and average response time, it should be median and mean in this scenario as median is a classification of average

Author: I agree but kept it as it is for clarity as this is what locust reports and what we have used in past reports.

|                          | 50% | 66% | 75% | 80% | 90% | 95% | 98% | 99% | 99.90% | 99.99% | 100% |
|--------------------------|-----|-----|-----|-----|-----|-----|-----|-----|--------|--------|------|
| Gunicorn Async           | 52     | 58  | 61     | 64  | 72  | 82  | 97  | 120 | 200    | 350    | 720  |
| Gunicorn Threads         | 60     | 67  | 72  | 77  | 92     | 110 | 140 | 160 | 250    | 450    | 1500 |
| Gunicorn Threads + gRPC  | 50     | 55  | 59     | 62  | 72  | 83  | 100 | 120 | 190   | 350   | 770  |

|                          | GETs Response Time | POSTs Response Time| All requests Response Time | 
|--------------------------|--------------------|--------------------|----------------------------|
| Gunicorn Async           | 87                 | 111                | 99                        |
| Gunicorn Threads         | 137                | 162                | 150                        |
| Gunicorn Threads + gRPC  | 91                 | 110                | 100                         |

> Performance comparison of Async vs Thread with gRPC over multiple tests.

Reviewer 2: - It's not clear what this means, are the numbers above not for multiple tests? 
- If this is associated with the image, think it would be clearer as a heading

Author: Made it a heading.

> I didn't include threads without gRPC since it was not performant enough to be considered.

Reviewer 2: "I didn't" doesn't fit with the rest of the document. It should be something like:
```suggestion
Threads without gRPC is not included as it was not performant enough.
```

![](outputs/performance_comparison.png)

## Observations

> - Threads with gRPC is just as performant as async

Reviewer 2: Until we run a scale test, we can't say that threads+grpc is as performant as async. Changing `is` to `was` focusses the observation on what was tested.

```suggestion
- Threads with gRPC was as performant as async
```

- Using threads is only feasible when also using GRPC

## Decision
- Given the performance of Gunicorn threads with GRPC is on par with async gevent and the options it opens up, we should use threads with GRPC

> - Look at the impact of varying the number of threads

Reviewer: Probably warrants an action heading

Author: Can change but still also a decision?. Just trying to keep it consistent with other reports. Not to bothered either way.

> - Test uWSGI threads with gRPC

Reviewer 2: We should add another action to run a scale test.
